From 3bf709e2b7925bbd0c010a97cbf2343632050c3b Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:09:30 -0400
Subject: [virt] kvm/ppc: book3s hv - Implement architecture compatibility modes for POWER8

Message-id: <1410545655-205645-342-git-send-email-dzickus@redhat.com>
Patchwork-id: 94607
O-Subject: [RHEL7 PATCH 341/626] KVM: PPC: Book3S HV: Implement architecture compatibility modes for POWER8
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 5557ae0ec77c2b4b5bbce2883c0603054ab66e68
Author: Paul Mackerras <paulus@samba.org>
Date:   Wed Jan 8 21:25:24 2014 +1100

    KVM: PPC: Book3S HV: Implement architecture compatibility modes for POWER8

    This allows us to select architecture 2.05 (POWER6) or 2.06 (POWER7)
    compatibility modes on a POWER8 processor.  (Note that transactional
    memory is disabled for usermode if either or both of the PCR_TM_DIS
    and PCR_ARCH_206 bits are set.)

    Signed-off-by: Paul Mackerras <paulus@samba.org>
    Signed-off-by: Alexander Graf <agraf@suse.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/reg.h b/arch/powerpc/include/asm/reg.h
index 52f74a4..d5495cd 100644
--- a/arch/powerpc/include/asm/reg.h
+++ b/arch/powerpc/include/asm/reg.h
@@ -330,6 +330,8 @@
 #define SPRN_PCR 0x152 /* Processor compatibility register */
 #define   PCR_VEC_DIS (1ul << (63-0)) /* Vec. disable (bit NA since POWER8) */
 #define   PCR_VSX_DIS (1ul << (63-1)) /* VSX disable (bit NA since POWER8) */
+#define   PCR_TM_DIS (1ul << (63-2)) /* Trans. memory disable (POWER8) */
+#define   PCR_ARCH_206 0x4  /* Architecture 2.06 */
 #define   PCR_ARCH_205 0x2  /* Architecture 2.05 */
 #define SPRN_HEIR 0x153 /* Hypervisor Emulated Instruction Register */
 #define SPRN_TLBINDEXR 0x154 /* P7 TLB control register */
diff --git a/arch/powerpc/kvm/book3s_hv.c b/arch/powerpc/kvm/book3s_hv.c
index 5146188..1045821 100644
--- a/arch/powerpc/kvm/book3s_hv.c
+++ b/arch/powerpc/kvm/book3s_hv.c
@@ -182,14 +182,28 @@ int kvmppc_set_arch_compat(struct kvm_vcpu *vcpu, u32 arch_compat)
 
   switch (arch_compat) {
   case PVR_ARCH_205:
-   pcr = PCR_ARCH_205;
+   /*
+    * If an arch bit is set in PCR, all the defined
+    * higher-order arch bits also have to be set.
+    */
+   pcr = PCR_ARCH_206 | PCR_ARCH_205;
    break;
   case PVR_ARCH_206:
   case PVR_ARCH_206p:
+   pcr = PCR_ARCH_206;
+   break;
+  case PVR_ARCH_207:
    break;
   default:
    return -EINVAL;
   }
+
+  if (!cpu_has_feature(CPU_FTR_ARCH_207S)) {
+   /* POWER7 can't emulate POWER8 */
+   if (!(pcr & PCR_ARCH_206))
+    return -EINVAL;
+   pcr &= ~PCR_ARCH_206;
+  }
  }
 
  spin_lock(&vc->lock);
-- 
1.7.1